highlight-wrapper/file-5.46/doc/magic.4
.\" $File: magic.man,v 1.110 2024/11/27 15:37:00 christos Exp $
.Dd November 27, 2024
.Dt MAGIC 4
.Os
.\" install as magic.4 on USG, magic.5 on V7, Berkeley and Linux systems.
.Sh NAME
.Nm magic
.Nd file command's magic pattern file
.Sh DESCRIPTION
This manual page documents the format of magic files as
used by the
.Xr file 1
command, version 5.46.
The
.Xr file 1
command identifies the type of a file using,
among other tests,
a test for whether the file contains certain
.Dq "magic patterns" .
The database of these
.Dq "magic patterns"
is usually located in a binary file in
.Pa /usr/local/share/misc/magic.mgc
or a directory of source text magic pattern fragment files in
.Pa /usr/local/share/misc/magic .
The database specifies what patterns are to be tested for, what message or
MIME type to print if a particular pattern is found,
and additional information to extract from the file.
.Pp
The format of the source fragment files that are used to build this database
is as follows:
Each line of a fragment file specifies a test to be performed.
A test compares the data starting at a particular offset
in the file with a byte value, a string or a numeric value.
If the test succeeds, a message is printed.
The line consists of the following fields:
.Bl -tag -width ".Dv message"
.It Dv offset
A number specifying the offset (in bytes) into the file of the data
which is to be tested.
This offset can be a negative number if it is:
.Bl -bullet  -compact
.It
The first direct offset of the magic entry (at continuation level 0),
in which case it is interpreted an offset from end end of the file
going backwards.
This works only when a file descriptor to the file is available and it
is a regular file.
.It
A continuation offset relative to the end of the last up-level field
.Dv ( \*[Am] ) .
.El
If the offset starts with the symbol
.Dq + ,
then all offsets are interpreted as from the beginning of the file (the
default).
.It Dv type
The type of the data to be tested.
The possible values are:
.Bl -tag -width ".Dv lestring16"
.It Dv byte
A one-byte value.
.It Dv short
A two-byte value in this machine's native byte order.
.It Dv long
A four-byte value in this machine's native byte order.
.It Dv quad
An eight-byte value in this machine's native byte order.
.It Dv float
A 32-bit single precision IEEE floating point number in this machine's native byte order.
.It Dv double
A 64-bit double precision IEEE floating point number in this machine's native byte order.
.It Dv string
A string of bytes.
The string type specification can be optionally followed by a /<width> 
option and optionally followed by a set of flags /[bCcftTtWw]*.
The width limits the number of characters to be copied.
Zero means all characters.
The following flags are supported:
.Bl -tag -width B -compact -offset XXXX
.It b
Force binary file test.
.It C
Use upper case insensitive matching: upper case
characters in the magic match both lower and upper case characters in the
target, whereas lower case characters in the magic only match upper case
characters in the target.
.It c
Use lower case insensitive matching: lower case
characters in the magic match both lower and upper case characters in the
target, whereas upper case characters in the magic only match upper case
characters in the target.
To do a complete case insensitive match, specify both
.Dq c
and
.Dq C .
.It f
Require that the matched string is a full word, not a partial word match.
.It T
Trim the string, i.e. leading and trailing whitespace
.It t
Force text file test.
.It W
Compact whitespace in the target, which must
contain at least one whitespace character.
If the magic has
.Dv n
consecutive blanks, the target needs at least
.Dv n
consecutive blanks to match.
.It w
Treat every blank in the magic as an optional blank.
is deleted before the string is printed.
.El
.It Dv pstring
A Pascal-style string where the first byte/short/int is interpreted as the
unsigned length.
The length defaults to byte and can be specified as a modifier.
The following modifiers are supported:
.Bl -tag -width B -compact -offset XXXX
.It B
A byte length (default).
.It H
A 2 byte big endian length.
.It h
A 2 byte little endian length.
.It L
A 4 byte big endian length.
.It l
A 4 byte little endian length.
.It J
The length includes itself in its count.
.El
The string is not NUL terminated.
.Dq J
is used rather than the more
valuable
.Dq I
because this type of length is a feature of the JPEG
format.
.It Dv date
A four-byte value interpreted as a UNIX date.
.It Dv qdate
An eight-byte value interpreted as a UNIX date.
.It Dv ldate
A four-byte value interpreted as a UNIX-style date, but interpreted as
local time rather than UTC.
.It Dv qldate
An eight-byte value interpreted as a UNIX-style date, but interpreted as
local time rather than UTC.
.It Dv qwdate
An eight-byte value interpreted as a Windows-style date.
.It Dv msdosdate
A two-byte value interpreted as FAT/DOS-style date.
.It Dv msdostime
A two-byte value interpreted as FAT/DOS-style time.
.It Dv beid3
A 32-bit ID3 length in big-endian byte order.
.It Dv beshort
A two-byte value in big-endian byte order.
.It Dv belong
A four-byte value in big-endian byte order.
.It Dv bequad
An eight-byte value in big-endian byte order.
.It Dv befloat
A 32-bit single precision IEEE floating point number in big-endian byte order.
.It Dv bedouble
A 64-bit double precision IEEE floating point number in big-endian byte order.
.It Dv bedate
A four-byte value in big-endian byte order,
interpreted as a Unix date.
.It Dv beqdate
An eight-byte value in big-endian byte order,
interpreted as a Unix date.
.It Dv beldate
A four-byte value in big-endian byte order,
interpreted as a UNIX-style date, but interpreted as local time rather
than UTC.
.It Dv beqldate
An eight-byte value in big-endian byte order,
interpreted as a UNIX-style date, but interpreted as local time rather
than UTC.
.It Dv beqwdate
An eight-byte value in big-endian byte order,
interpreted as a Windows-style date.
.It Dv bemsdosdate
A two-byte value in big-endian byte order,
interpreted as FAT/DOS-style date.
.It Dv bemsdostime
A two-byte value in big-endian byte order,
interpreted as FAT/DOS-style time.
.It Dv bestring16
A two-byte unicode (UCS16) string in big-endian byte order.
.It Dv leid3
A 32-bit ID3 length in little-endian byte order.
.It Dv leshort
A two-byte value in little-endian byte order.
.It Dv lelong
A four-byte value in little-endian byte order.
.It Dv lequad
An eight-byte value in little-endian byte order.
.It Dv lefloat
A 32-bit single precision IEEE floating point number in little-endian byte order.
.It Dv ledouble
A 64-bit double precision IEEE floating point number in little-endian byte order.
.It Dv ledate
A four-byte value in little-endian byte order,
interpreted as a UNIX date.
.It Dv leqdate
An eight-byte value in little-endian byte order,
interpreted as a UNIX date.
.It Dv leldate
A four-byte value in little-endian byte order,
interpreted as a UNIX-style date, but interpreted as local time rather
than UTC.
.It Dv leqldate
An eight-byte value in little-endian byte order,
interpreted as a UNIX-style date, but interpreted as local time rather
than UTC.
.It Dv leqwdate
An eight-byte value in little-endian byte order,
interpreted as a Windows-style date.
.It Dv lemsdosdate
A two-byte value in big-endian byte order,
interpreted as FAT/DOS-style date.
.It Dv lemsdostime
A two-byte value in big-endian byte order,
interpreted as FAT/DOS-style time.
.It Dv lestring16
A two-byte unicode (UCS16) string in little-endian byte order.
.It Dv melong
A four-byte value in middle-endian (PDP-11) byte order.
.It Dv medate
A four-byte value in middle-endian (PDP-11) byte order,
interpreted as a UNIX date.
.It Dv meldate
A four-byte value in middle-endian (PDP-11) byte order,
interpreted as a UNIX-style date, but interpreted as local time rather
than UTC.
.It Dv indirect
Starting at the given offset, consult the magic database again.
The offset of the
.Dv indirect
magic is by default absolute in the file, but one can specify
.Dv /r
to indicate that the offset is relative from the beginning of the entry.
.It Dv name
Define a
.Dq named
magic instance that can be called from another
.Dv use
magic entry, like a subroutine call.
Named instance direct magic offsets are relative to the offset of the
previous matched entry, but indirect offsets are relative to the beginning
of the file as usual.
Named magic entries always match.
.It Dv use
Recursively call the named magic starting from the current offset.
If the name of the referenced begins with a
.Dv ^
then the endianness of the magic is switched; if the magic mentioned
.Dv leshort
for example,
it is treated as
.Dv beshort
and vice versa.
This is useful to avoid duplicating the rules for different endianness.
.It Dv regex
A regular expression match in extended POSIX regular expression syntax
(like egrep).
Regular expressions can take exponential time to process, and their
performance is hard to predict, so their use is discouraged.
When used in production environments, their performance
should be carefully checked.
The size of the string to search should also be limited by specifying
.Dv /<length> ,
to avoid performance issues scanning long files.
The type specification can also be optionally followed by
.Dv /[c][s][l] .
The
.Dq c
flag makes the match case insensitive, while the
.Dq s
flag update the offset to the start offset of the match, rather than the end.
The
.Dq l
modifier, changes the limit of length to mean number of lines instead of a
byte count.
Lines are delimited by the platforms native line delimiter.
When a line count is specified, an implicit byte count also computed assuming
each line is 80 characters long.
If neither a byte or line count is specified, the search is limited automatically
to 8KiB.
.Dv ^
and
.Dv $
match the beginning and end of individual lines, respectively,
not beginning and end of file.
.It Dv search
A literal string search starting at the given offset.
The same modifier flags can be used as for string patterns.
The search expression must contain the range in the form
.Dv /number,
that is the number of positions at which the match will be
attempted, starting from the start offset.
This is suitable for
searching larger binary expressions with variable offsets, using
.Dv \e

CHANGELOG.md
# Changelog


### 1.0.23 (49)
Bugfix:
- Unable to save the settings.


### 1.0.22 (47)
New Features:
- Support for MDX files (`.mdx`, _without_ JSX rendering).
- Support for Cursor Rulers files (`.mdc`).
- Highlight updated to 4.16.
- Lua updated to 5.4.7.
- File updated to 5.46.
- Enry updated to 2.9.2.
- Support for UTI `io.typora.markdown`.

Bugfix:
- Rewrite settings management to bypass warnings about accessing settings from other applications. **You will need to reset any preferences you have changed.**
- Unused code has been removed (including the ability to choose a theme for code formatting, now to change it you will need to edit the css file)


### 1.0.21 (46)
New Features:
- Warning about this is not a Markdown editor app.
- Alert when the app is closed and the current markdown file was changed.
- Experimental Shortcut Action (require macOS 15.2). 

Bugfix:
- Temporary fix for $ recognition without mathematical expression.


### 1.0.20 (45)
New Features:
- Support for some extended syntax:
  - highlight text with `==`
  - subscript (`~`) and superscript (`^`).


### 1.0.19 (44)
Bugfix:
- Fixed saving about flag settings.


### 1.0.18 (43)
New Features:
- Support for `com.nutstore.down` UTI.
- cmark-gfm updated to version 0.29.0.gfm.13.
- pcre2 updated to version 10.43.
- Enry updated to version 2.8.8.
- Settings to show about app info on the footer preview.
- Link to buy me a :coffee: (:heart:)

Bugfix:
- Exported html head title filled with the filename.


### 1.0.17 (42)
New Features:
- Improved Math (Latex) support (thanks to @mccann):
    - Prevent Markdown processing inside $'s
    - Allow multi-line Latex, like that supported by:
        - jupyter notebook
        - vscode
        - obsidian
        - probably many more - only tested the above.
    - Now prevents arbitrary html getting passed through.


### 1.0.16 (41)
New features:
- Support mathematical expressions with the MathJax engine.
- Update libraries:
    - cmark-gfm 0.29.0.gfm.11
    - highlight 4.5
    - lua 5.4.4
    - libmagic 5.44
    - go-enry v2.8.4
    - boost.
- Settings UI redesigned.
  
Bugfix:
- Alt attribute for emoji rendered as image.
 

### 1.0.15 (40)
New features:
- The markdown file is automatically reloaded when it is edited outside of the app.  
- Preliminary support for Quarto files (`.qmd`).


### 1.0.14 (39)
Bugfix:
- Allows you to use `...` to end the `yaml` header block. 


### 1.0.13 (38)
Bugfix:
- Fixed the Sparkle integration bug. **If you have installed version 1.0.11 or 1.0.12 you may need to [re-download the updated app from the web](https://github.com/sbarex/QLMarkdown/releases/download/1.0.13/QLMarkdown.zip).** 


### 1.0.12 (37)
Bugfix:
- Better procedure to install the command line tool.
- Fixed the bug that prevented the processing of html raw images when they are not inserted in a html block element. 


### 1.0.11 (36)
New features:
- Support for opening markdown files (by dragging the file onto the app icon)
- Support for exporting the markdown code.
- Sparkle updated to release 2.0.0. 

Bugfix:
- Fix for heads with dash.
- Implemented missing behavior for the color scheme editor.
- Fix for installation of the command line tool.


### 1.0.10 (35)
New features:
- Better performance for heads extension.
- Better performance for inline images on raw html fragments.
- Option for automatic saving of settings changes.
- GUI optimization.

Bugfix:
- Fixed settings save.


### 1.0.9 (34)
New features:
- Experimental support for defining the size of the Quick Look window.
- Magic `file` updated to release 5.41.

Bugfix:
- Fixed cross compile compilation of `file` library.


### 1.0.8 (33)
Bugfix:
- Fixed bug for undefined source code theme.
- Fixed bug on settings sync delay in the Quick Look preview on macOS Monterey.


### 1.0.7 (32)
New features:
- Support for render the markdown source code instead of the formatted output.


### 1.0.6 (31)
New features:
- Application menu item to install/reveal the CLI tool on `/usr/local/bin` folder.
- Support for UTI `public.markdown` defined by `Nova.app`.

Bugfix:
- Fixed possibile deallocation error with invalid highlight theme.


### 1.0.5 (30)
Bugfix:
- Tag Filter extension fixed.
- Improved performance for inline image handling in raw HTML fragment.
- Better recognition of javascript fenced block.
- Missing library credits on about dialog.


### 1.0.4 (29)
New features:
- On macOS 12 Monterey, the new lightweight data based preview will be used.
- Addeded more arguments to the CLI tool.
- CLI tool embed inline images handle also raw HTML fragments.

Bugfix:
- Handle markdown associated only to a dynamic UTI.
- Fixed issue with instantiated and never released `Markdown QL Extension Web Content` process for each file previewed.
- On macOS 12 Monterey now you can scroll the preview with the scrollbars.
- On macOS 12 Monterey the bug of trackpad scrolling in fullscreen preview has been fixed.


### 1.0.3 (28)
New features:
- Command line interface (CLI) tool.


### 1.0.2 (27)
Bugfix:
- Fenced code block highlighted even when rmd syntax is used (language name inside a curly brace).
  

### 1.0.1 (26)

highlight-wrapper/file-5.44/src/magic.c
/*
 * Copyright (c) Christos Zoulas 2003.
 * All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice immediately at the beginning of the file, without modification,
 *    this list of conditions, and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifdef WIN32
#include <windows.h>
#include <shlwapi.h>
#endif

#include "file.h"

#ifndef	lint
FILE_RCSID("@(#)$File: magic.c,v 1.120 2022/12/26 17:31:14 christos Exp $")
#endif	/* lint */

#include "magic.h"

#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#ifdef QUICK
#include <sys/mman.h>
#endif
#include <limits.h>	/* for PIPE_BUF */

#if defined(HAVE_UTIMES)
# include <sys/time.h>
#elif defined(HAVE_UTIME)
# if defined(HAVE_SYS_UTIME_H)
#  include <sys/utime.h>
# elif defined(HAVE_UTIME_H)
#  include <utime.h>
# endif
#endif

#ifdef HAVE_UNISTD_H
#include <unistd.h>	/* for read() */
#endif

#ifndef PIPE_BUF
/* Get the PIPE_BUF from pathconf */
#ifdef _PC_PIPE_BUF
#define PIPE_BUF pathconf(".", _PC_PIPE_BUF)
#else
#define PIPE_BUF 512
#endif
#endif

file_private void close_and_restore(const struct magic_set *, const char *, int,
    const struct stat *);
file_private int unreadable_info(struct magic_set *, mode_t, const char *);
file_private const char* get_default_magic(void);
#ifndef COMPILE_ONLY
file_private const char *file_or_fd(struct magic_set *, const char *, int);
#endif

#ifndef	STDIN_FILENO
#define	STDIN_FILENO	0
#endif

#ifdef WIN32
/* HINSTANCE of this shared library. Needed for get_default_magic() */
static HINSTANCE _w32_dll_instance = NULL;

static void
_w32_append_path(char **hmagicpath, const char *fmt, ...)
{
	char *tmppath;
        char *newpath;
	va_list ap;

	va_start(ap, fmt);
	if (vasprintf(&tmppath, fmt, ap) < 0) {
		va_end(ap);
		return;
	}
	va_end(ap);

	if (access(tmppath, R_OK) == -1)
		goto out;

	if (*hmagicpath == NULL) {
		*hmagicpath = tmppath;
		return;
	}

	if (asprintf(&newpath, "%s%c%s", *hmagicpath, PATHSEP, tmppath) < 0)
		goto out;

	free(*hmagicpath);
	free(tmppath);
	*hmagicpath = newpath;
	return;
out:
	free(tmppath);
}

static void
_w32_get_magic_relative_to(char **hmagicpath, HINSTANCE module)
{
	static const char *trypaths[] = {
		"%s/share/misc/magic.mgc",
		"%s/magic.mgc",
	};
	LPSTR dllpath;
	size_t sp;

	dllpath = calloc(MAX_PATH + 1, sizeof(*dllpath));

	if (!GetModuleFileNameA(module, dllpath, MAX_PATH))
		goto out;

	PathRemoveFileSpecA(dllpath);

	if (module) {
		char exepath[MAX_PATH];
		GetModuleFileNameA(NULL, exepath, MAX_PATH);
		PathRemoveFileSpecA(exepath);
		if (stricmp(exepath, dllpath) == 0)
			goto out;
	}

	sp = strlen(dllpath);
	if (sp > 3 && stricmp(&dllpath[sp - 3], "bin") == 0) {
		_w32_append_path(hmagicpath,
		    "%s/../share/misc/magic.mgc", dllpath);
		goto out;
	}

	for (sp = 0; sp < __arraycount(trypaths); sp++)
		_w32_append_path(hmagicpath, trypaths[sp], dllpath);
out:
	free(dllpath);
}

#ifndef BUILD_AS_WINDOWS_STATIC_LIBARAY
/* Placate GCC by offering a sacrificial previous prototype */
BOOL WINAPI DllMain(HINSTANCE, DWORD, LPVOID);

BOOL WINAPI
DllMain(HINSTANCE hinstDLL, DWORD fdwReason,
    LPVOID lpvReserved __attribute__((__unused__)))
{
	if (fdwReason == DLL_PROCESS_ATTACH)
		_w32_dll_instance = hinstDLL;
	return 1;
}
#endif
#endif

file_private const char *
get_default_magic(void)
{
	static const char hmagic[] = "/.magic/magic.mgc";
	static char *default_magic;
	char *home, *hmagicpath;

#ifndef WIN32
	struct stat st;

	if (default_magic) {
		free(default_magic);
		default_magic = NULL;
	}
	if ((home = getenv("HOME")) == NULL)
		return MAGIC;

	if (asprintf(&hmagicpath, "%s/.magic.mgc", home) < 0)
		return MAGIC;
	if (stat(hmagicpath, &st) == -1) {
		free(hmagicpath);
		if (asprintf(&hmagicpath, "%s/.magic", home) < 0)
			return MAGIC;
		if (stat(hmagicpath, &st) == -1)
			goto out;
		if (S_ISDIR(st.st_mode)) {
			free(hmagicpath);
			if (asprintf(&hmagicpath, "%s/%s", home, hmagic) < 0)
				return MAGIC;
			if (access(hmagicpath, R_OK) == -1)
				goto out;
		}
	}

	if (asprintf(&default_magic, "%s:%s", hmagicpath, MAGIC) < 0)
		goto out;
	free(hmagicpath);
	return default_magic;
out:
	default_magic = NULL;
	free(hmagicpath);
	return MAGIC;
#else
	hmagicpath = NULL;

	if (default_magic) {
		free(default_magic);
		default_magic = NULL;
	}

	/* Before anything else, try to get a magic file from user HOME */
	if ((home = getenv("HOME")) != NULL)
		_w32_append_path(&hmagicpath, "%s%s", home, hmagic);

	/* First, try to get a magic file from user-application data */
	if ((home = getenv("LOCALAPPDATA")) != NULL)
		_w32_append_path(&hmagicpath, "%s%s", home, hmagic);

	/* Second, try to get a magic file from the user profile data */
	if ((home = getenv("USERPROFILE")) != NULL)
		_w32_append_path(&hmagicpath,
		    "%s/Local Settings/Application Data%s", home, hmagic);

	/* Third, try to get a magic file from Common Files */
	if ((home = getenv("COMMONPROGRAMFILES")) != NULL)
		_w32_append_path(&hmagicpath, "%s%s", home, hmagic);

	/* Fourth, try to get magic file relative to exe location */
        _w32_get_magic_relative_to(&hmagicpath, NULL);

	/* Fifth, try to get magic file relative to dll location */
        _w32_get_magic_relative_to(&hmagicpath, _w32_dll_instance);

	/* Avoid MAGIC constant - it likely points to a file within MSys tree */
	default_magic = hmagicpath;
	return default_magic;
#endif
}

file_public const char *
magic_getpath(const char *magicfile, int action)
{
	if (magicfile != NULL)
		return magicfile;

	magicfile = getenv("MAGIC");
	if (magicfile != NULL)
		return magicfile;

	return action == FILE_LOAD ? get_default_magic() : MAGIC;
}

file_public struct magic_set *
magic_open(int flags)
{
	return file_ms_alloc(flags);
}

file_private int
unreadable_info(struct magic_set *ms, mode_t md, const char *file)
{
	if (file) {
		/* We cannot open it, but we were able to stat it. */
		if (access(file, W_OK) == 0)
			if (file_printf(ms, "writable, ") == -1)
				return -1;
#ifndef WIN32
		if (access(file, X_OK) == 0)
			if (file_printf(ms, "executable, ") == -1)
				return -1;
#else
		/* X_OK doesn't work well on MS-Windows */
		{
			const char *p = strrchr(file, '.');
			if (p && (stricmp(p, ".exe")
				  || stricmp(p, ".dll")
				  || stricmp(p, ".bat")
				  || stricmp(p, ".cmd")))
				if (file_printf(ms, "writable, ") == -1)
					return -1;
		}
#endif
	}
	if (S_ISREG(md))
		if (file_printf(ms, "regular file, ") == -1)
			return -1;
	if (file_printf(ms, "no read permission") == -1)
		return -1;
	return 0;
}

file_public void
magic_close(struct magic_set *ms)
{
	if (ms == NULL)
		return;
	file_ms_free(ms);
}

/*
 * load a magic file
 */
file_public int
magic_load(struct magic_set *ms, const char *magicfile)
{
	if (ms == NULL)
		return -1;
	return file_apprentice(ms, magicfile, FILE_LOAD);
}

#ifndef COMPILE_ONLY
/*
 * Install a set of compiled magic buffers.
 */
file_public int
magic_load_buffers(struct magic_set *ms, void **bufs, size_t *sizes,
    size_t nbufs)
{
	if (ms == NULL)
		return -1;
	return buffer_apprentice(ms, RCAST(struct magic **, bufs),
	    sizes, nbufs);
}
#endif

file_public int
magic_compile(struct magic_set *ms, const char *magicfile)
{
	if (ms == NULL)
		return -1;
	return file_apprentice(ms, magicfile, FILE_COMPILE);
}

file_public int
magic_check(struct magic_set *ms, const char *magicfile)
{
	if (ms == NULL)
		return -1;
	return file_apprentice(ms, magicfile, FILE_CHECK);
}

file_public int
magic_list(struct magic_set *ms, const char *magicfile)
{
	if (ms == NULL)
		return -1;
	return file_apprentice(ms, magicfile, FILE_LIST);
}

file_private void
close_and_restore(const struct magic_set *ms, const char *name, int fd,
    const struct stat *sb)
{
	if (fd == STDIN_FILENO || name == NULL)
		return;
	(void) close(fd);

	if ((ms->flags & MAGIC_PRESERVE_ATIME) != 0) {
		/*
		 * Try to restore access, modification times if read it.
		 * This is really *bad* because it will modify the status
		 * time of the file... And of course this will affect
		 * backup programs
		 */
#ifdef HAVE_UTIMES
		struct timeval  utsbuf[2];
		(void)memset(utsbuf, 0, sizeof(utsbuf));
		utsbuf[0].tv_sec = sb->st_atime;
		utsbuf[1].tv_sec = sb->st_mtime;

		(void) utimes(name, utsbuf); /* don't care if loses */
#elif defined(HAVE_UTIME_H) || defined(HAVE_SYS_UTIME_H)
		struct utimbuf  utbuf;

		(void)memset(&utbuf, 0, sizeof(utbuf));
		utbuf.actime = sb->st_atime;
		utbuf.modtime = sb->st_mtime;
		(void) utime(name, &utbuf); /* don't care if loses */
#endif
	}
}

#ifndef COMPILE_ONLY

/*
 * find type of descriptor
 */
file_public const char *
magic_descriptor(struct magic_set *ms, int fd)
{
	if (ms == NULL)
		return NULL;
	return file_or_fd(ms, NULL, fd);
}

/*
 * find type of named file
 */
file_public const char *
magic_file(struct magic_set *ms, const char *inname)
{
	if (ms == NULL)
		return NULL;
	return file_or_fd(ms, inname, STDIN_FILENO);
}

file_private const char *
file_or_fd(struct magic_set *ms, const char *inname, int fd)
{
	int	rv = -1;
	unsigned char *buf;
	struct stat	sb;
	ssize_t nbytes = 0;	/* number of bytes read from a datafile */
	int	ispipe = 0;
	int	okstat = 0;
	off_t	pos = CAST(off_t, -1);

	if (file_reset(ms, 1) == -1)
		goto out;

	/*
	 * one extra for terminating '\0', and
	 * some overlapping space for matches near EOF
	 */
#define SLOP (1 + sizeof(union VALUETYPE))
	if ((buf = CAST(unsigned char *, malloc(ms->bytes_max + SLOP))) == NULL)
		return NULL;

	switch (file_fsmagic(ms, inname, &sb)) {
	case -1:		/* error */
		goto done;
	case 0:			/* nothing found */
		break;
	default:		/* matched it and printed type */
		rv = 0;
		goto done;
	}

#ifdef WIN32
	/* Place stdin in binary mode, so EOF (Ctrl+Z) doesn't stop early. */
	if (fd == STDIN_FILENO)
		_setmode(STDIN_FILENO, O_BINARY);
#endif
	if (inname != NULL) {
		int flags = O_RDONLY|O_BINARY|O_NONBLOCK|O_CLOEXEC;
		errno = 0;
		if ((fd = open(inname, flags)) < 0) {
			okstat = stat(inname, &sb) == 0;
#ifdef WIN32
			/*
			 * Can't stat, can't open.  It may have been opened in
			 * fsmagic, so if the user doesn't have read permission,
			 * allow it to say so; otherwise an error was probably
			 * displayed in fsmagic.
			 */
			if (!okstat && errno == EACCES) {
				sb.st_mode = S_IFBLK;
				okstat = 1;
			}
#endif
			if (okstat &&
			    unreadable_info(ms, sb.st_mode, inname) == -1)
				goto done;
			rv = 0;
			goto done;
		}
#if O_CLOEXEC == 0 && defined(F_SETFD)
		(void)fcntl(fd, F_SETFD, FD_CLOEXEC);
#endif
	}

	if (fd != -1) {
		okstat = fstat(fd, &sb) == 0;
		if (okstat && S_ISFIFO(sb.st_mode))
			ispipe = 1;
		if (inname == NULL)
			pos = lseek(fd, CAST(off_t, 0), SEEK_CUR);
	}

	/*
	 * try looking at the first ms->bytes_max bytes
	 */
	if (ispipe) {
		if (fd != -1) {
			ssize_t r = 0;

			while ((r = sread(fd, RCAST(void *, &buf[nbytes]),
			    CAST(size_t, ms->bytes_max - nbytes), 1)) > 0) {
				nbytes += r;
				if (r < PIPE_BUF) break;
			}
		}

		if (nbytes == 0 && inname) {
			/* We can not read it, but we were able to stat it. */
			if (unreadable_info(ms, sb.st_mode, inname) == -1)
				goto done;
			rv = 0;
			goto done;
		}

	} else if (fd != -1) {
		/* Windows refuses to read from a big console buffer. */
		size_t howmany =
#ifdef WIN32
		    _isatty(fd) ? 8 * 1024 :
#endif
		    ms->bytes_max;
		if ((nbytes = read(fd, RCAST(void *, buf), howmany)) == -1) {
			if (inname == NULL && fd != STDIN_FILENO)
				file_error(ms, errno, "cannot read fd %d", fd);
			else
				file_error(ms, errno, "cannot read `%s'",
				    inname == NULL ? "/dev/stdin" : inname);
			goto done;
		}
	}

	(void)memset(buf + nbytes, 0, SLOP); /* NUL terminate */
	if (file_buffer(ms, fd, okstat ? &sb : NULL, inname, buf, CAST(size_t, nbytes)) == -1)
		goto done;
	rv = 0;
done:
	free(buf);
	if (fd != -1) {
		if (pos != CAST(off_t, -1))
			(void)lseek(fd, pos, SEEK_SET);
		close_and_restore(ms, inname, fd, &sb);
	}
out:
	return rv == 0 ? file_getbuffer(ms) : NULL;
}


file_public const char *
magic_buffer(struct magic_set *ms, const void *buf, size_t nb)
{
	if (ms == NULL)
		return NULL;
	if (file_reset(ms, 1) == -1)
		return NULL;
	/*
	 * The main work is done here!
	 * We have the file name and/or the data buffer to be identified.
	 */
	if (file_buffer(ms, -1, NULL, NULL, buf, nb) == -1) {
		return NULL;
	}
	return file_getbuffer(ms);
}
#endif

file_public const char *
magic_error(struct magic_set *ms)
{
	if (ms == NULL)
		return "Magic database is not open";
	return (ms->event_flags & EVENT_HAD_ERR) ? ms->o.buf : NULL;
}

file_public int
magic_errno(struct magic_set *ms)
{
	if (ms == NULL)
		return EINVAL;
	return (ms->event_flags & EVENT_HAD_ERR) ? ms->error : 0;
}

file_public int
magic_getflags(struct magic_set *ms)
{
	if (ms == NULL)
		return -1;

	return ms->flags;
}

file_public int
magic_setflags(struct magic_set *ms, int flags)
{
	if (ms == NULL)
		return -1;
#if !defined(HAVE_UTIME) && !defined(HAVE_UTIMES)
	if (flags & MAGIC_PRESERVE_ATIME)
		return -1;
#endif
	ms->flags = flags;
	return 0;
}

file_public int
magic_version(void)
{
	return MAGIC_VERSION;
}

file_public int
magic_setparam(struct magic_set *ms, int param, const void *val)
{
	if (ms == NULL)
		return -1;
	switch (param) {
	case MAGIC_PARAM_INDIR_MAX:
		ms->indir_max = CAST(uint16_t, *CAST(const size_t *, val));
		return 0;
	case MAGIC_PARAM_NAME_MAX:
		ms->name_max = CAST(uint16_t, *CAST(const size_t *, val));
		return 0;
	case MAGIC_PARAM_ELF_PHNUM_MAX:
		ms->elf_phnum_max = CAST(uint16_t, *CAST(const size_t *, val));
		return 0;
	case MAGIC_PARAM_ELF_SHNUM_MAX:
		ms->elf_shnum_max = CAST(uint16_t, *CAST(const size_t *, val));
		return 0;
	case MAGIC_PARAM_ELF_NOTES_MAX:
		ms->elf_notes_max = CAST(uint16_t, *CAST(const size_t *, val));
		return 0;
	case MAGIC_PARAM_REGEX_MAX:
		ms->regex_max = CAST(uint16_t, *CAST(const size_t *, val));
		return 0;
	case MAGIC_PARAM_BYTES_MAX:
		ms->bytes_max = *CAST(const size_t *, val);
		return 0;
	case MAGIC_PARAM_ENCODING_MAX:
		ms->encoding_max = *CAST(const size_t *, val);
		return 0;
	default:
		errno = EINVAL;
		return -1;
	}
}

file_public int
magic_getparam(struct magic_set *ms, int param, void *val)
{
	if (ms == NULL)
		return -1;
	switch (param) {
	case MAGIC_PARAM_INDIR_MAX:
		*CAST(size_t *, val) = ms->indir_max;
		return 0;
	case MAGIC_PARAM_NAME_MAX:
		*CAST(size_t *, val) = ms->name_max;
		return 0;
	case MAGIC_PARAM_ELF_PHNUM_MAX:
		*CAST(size_t *, val) = ms->elf_phnum_max;
		return 0;
	case MAGIC_PARAM_ELF_SHNUM_MAX:
		*CAST(size_t *, val) = ms->elf_shnum_max;
		return 0;
	case MAGIC_PARAM_ELF_NOTES_MAX:
		*CAST(size_t *, val) = ms->elf_notes_max;
		return 0;
	case MAGIC_PARAM_REGEX_MAX:
		*CAST(size_t *, val) = ms->regex_max;
		return 0;
	case MAGIC_PARAM_BYTES_MAX:
		*CAST(size_t *, val) = ms->bytes_max;
		return 0;
	case MAGIC_PARAM_ENCODING_MAX:
		*CAST(size_t *, val) = ms->encoding_max;
		return 0;
	default:
		errno = EINVAL;
		return -1;
	}
}
